nixos/modules/services/web-apps/peering-manager.nix

```diff
@@ -171,9 +199,18 @@ in {
       '' + lib.optionalString (cfg.peeringdbApiKeyFile != null) ''
         with open("${cfg.peeringdbApiKeyFile}", "r") as file:
           PEERINGDB_API_KEY = file.readline()
+      '' + lib.optionalString (cfg.enableOidc && cfg.oidcSecretFile != null) ''
+        with open("${cfg.oidcSecretFile}", "r") as file:
+          OIDC_RP_CLIENT_SECRET = file.readline()
```

Comment:
I think this won't work, because it is added to the `configuration.py` and not to the `oidc_config.py`. But in my opinion we don't really need this, since when the OIDC config file is provided by something like SOPS, it (and especially the secret) will be encrypted and won't be in the world-readable nix-store.

Reply:
I need to test it, but I don't think it's nessesary that it's in `oidc_config.py`. The files should be evaluated in `settings.py` and then merged anyway:
```py
try:
    from peering_manager.oidc_config import *

    OIDC_CONFIGURED = True
except ImportError:
    OIDC_CONFIGURED = False
```

But thanks for the hint.

Reply:
This just imports everything from the `oidc_config.py`, but not the other way around. The settings from the `configuration.py` are imported selectively like this (just a small snippet):
```python
CORS_ORIGIN_ALLOW_ALL = getattr(configuration, "CORS_ORIGIN_ALLOW_ALL", False)
CORS_ORIGIN_REGEX_WHITELIST = getattr(configuration, "CORS_ORIGIN_REGEX_WHITELIST", [])
CORS_ORIGIN_WHITELIST = getattr(configuration, "CORS_ORIGIN_WHITELIST", [])
CSRF_COOKIE_NAME = getattr(configuration, "CSRF_COOKIE_NAME", "csrftoken")
CSRF_COOKIE_SECURE = getattr(configuration, "CSRF_COOKIE_SECURE", False)
CSRF_TRUSTED_ORIGINS = getattr(configuration, "CSRF_TRUSTED_ORIGINS", [])
DEBUG = getattr(configuration, "DEBUG", False)
LOGGING = getattr(configuration, "LOGGING", {})
REDIS = getattr(configuration, "REDIS", {})
```

And since `OIDC_RP_CLIENT_SECRET` isn't explicitly imported from the `configuration.py`, I think this won't work.

Reply:
Mhm ok, I need to test it :D I thought that it would be imported into the django settings and from there configure oidc, but I might be wrong.
```py
# Import configuration parameters
config_path = os.getenv("PEERINGMANAGER_CONFIGURATION", "peering_manager.configuration")
try:
    configuration = importlib.import_module(config_path)
except ModuleNotFoundError as e:
    if getattr(e, "name") == config_path:
        raise ImproperlyConfigured(
            f"Specified configuration module ({config_path}) not found. Please define peering_manager/configuration.py per the documentation, or specify an alternate module in the PEERINGMANAGER_CONFIGURATION environment variable."
        )
    raise

# ...

try:
    from peering_manager.oidc_config import *

    OIDC_CONFIGURED = True
except ImportError:
    OIDC_CONFIGURED = False


def generate_username(email):
    return unicodedata.normalize("NFKC", email)[:150]


OIDC_USERNAME_ALGO = generate_username

if OIDC_CONFIGURED:
    AUTHENTICATION_BACKENDS = [
        "mozilla_django_oidc.auth.OIDCAuthenticationBackend",
        "django.contrib.auth.backends.ModelBackend",
    ]
    INSTALLED_APPS.insert(2, "mozilla_django_oidc")
    MIDDLEWARE.insert(5, "mozilla_django_oidc.middleware.SessionRefresh")

```
If I would have build mozilla-django-oidc I would access the variable from the middleware using `django.conf.settings.OIDC_RP_CLIENT_SECRET` directly and not try to read it again... But as I said, I didn't check this yet.

Reply:
![image](https://github.com/NixOS/nixpkgs/assets/26925347/8e86281f-e114-4b3c-b431-72fdc5b965dd)
Okkkkkkay... Maybe you're right :) I will think about another solution

Reply:
First I tried adding it to the oidc-config.py using something like
```nix
        oidcConfigPath = (pkgs.writeText "oidc-config.py" ''
          # ...          
          with open("${sops.secrets."peering-manager/oidcSecret".path}", "r") as file:
              OIDC_RP_CLIENT_SECRET = file.readline()
        '');
```
which results also in errors, so I will have to implement it using sops templates:
```nix
    sops = {
      secrets = {
        "peering-manager/secretKey".owner = "peering-manager";
      } // (lib.optionalAttrs (config.my-service.peering-manager.oidc.domain != "") {
        "peering-manager/oidcSecret".owner = "peering-manager";
      });

      templates."peering-manager/oidc-config".content = ''
        # CLIENT_ID and SECRET are required to authenticate against the provider
        OIDC_RP_CLIENT_ID = "${config.my-service.peering-manager.oidc.clientId}"
        OIDC_RP_CLIENT_SECRET = ${config.sops.placeholder."peering-manager/oidcSecret"}

        # The following two may be required depending on your provider,
        # check the configuration endpoint for JWKS information
        OIDC_RP_SIGN_ALGO = "RS256";
        OIDC_OP_JWKS_ENDPOINT = "https://${config.my-service.peering-manager.oidc.domain}/realms/${config.my-service.peering-manager.oidc.realm}/protocol/openid-connect/certs"

        # Refer to the configuration endpoint of your provider
        OIDC_OP_AUTHORIZATION_ENDPOINT = "https://${config.my-service.peering-manager.oidc.domain}/realms/${config.my-service.peering-manager.oidc.realm}/protocol/openid-connect/auth"
        OIDC_OP_TOKEN_ENDPOINT = "https://${config.my-service.peering-manager.oidc.domain}/realms/${config.my-service.peering-manager.oidc.realm}/protocol/openid-connect/token"
        OIDC_OP_USER_ENDPOINT = "https://${config.my-service.peering-manager.oidc.domain}/realms/${config.my-service.peering-manager.oidc.realm}/protocol/openid-connect/userinfo"

        # Set these to the base path of your Peering Manager installation
        LOGIN_REDIRECT_URL = "https://${config.my-service.peering-manager.domain}"
        LOGOUT_REDIRECT_URL = "https://${config.my-service.peering-manager.domain}"

        # If this is True, new users will be created if not yet existing.
        OIDC_CREATE_USER = True
      '';
    };
    # ...
    services.peering-manager.oidcConfigPath = config.sops.templates."peering-manager/oidc-config".path;
```
which seams to work so far.